ui: Order submit button state management

peterzen opened the pull request:
This PR modifies the behavior of the order submit button so that it prevents the user from submitting the order if it would fail due to  the below conditions:

* quantity is 0
* buy or sell quantity is more than available balance 

Closes #2527

JoeGruffins commented:
Will also say insufficient balance if lots are set to zero but have balance.
![image](https://github.com/decred/dcrdex/assets/34998433/55600edd-ca11-4a53-9a12-8d9c10ba664f)

peterzen commented:
> Will also say insufficient balance if lots are set to zero but have balance. ![image](https://user-images.githubusercontent.com/34998433/270628892-55600edd-ca11-4a53-9a12-8d9c10ba664f.png)

Thanks for catching that, it's fixed.

martonp reviewed (commented):
For other reasons why a user would not be able to trade, such as the bond pending or one of the wallets not yet being created, we are removing the order form completely. See `resolveOrderFormVisibility` in `markets.ts`. Personally, I don't think the order form should be shown if the user has no balance in either of the assets. Why should they be entering a rate and # of lots if they cannot trade anyways? Also, if only one of the assets has a balance, then you can just grey out the Buy/Sell button on top that switches to the other side of the market.

Thoughts? @buck54321 @JoeGruffins ?

martonp commented:
For markets involving ETH/Tokens, if there is no ETH balance in the wallet, then the user will not be able to trade at all, even if they are buying ETH or the ETH token. The user needs ETH to redeem.

peterzen commented:
> Personally, I don't think the order form should be shown if the user has no balance in either of the assets. Why should they be entering a rate and # of lots if they cannot trade anyways? Also, if only one of the assets has a balance, then you can just grey out the Buy/Sell button on top that switches to the other side of the market.
> 
A new user will most likely want to click around to discover the UI/UX before depositing funds - if they can't find the order form anywhere that's unusual/confusing and will likely to generate support questions.

What problem would the hidden form solve?

martonp commented:
> What problem would the hidden form solve?

It would be more consistent with the rest of the features. If one of the wallets isn't created, then it will also not show the order form. I think if one of the assets doesn't have funds, then automatically showing the side of the market that they are able to use and not allowing them to click to the other side is the ideal UI. And as an extension of that, if they are not able to use either of the sides, then just a message telling them to deposit funds would be shown instead of the order form. I don't think anyone would complain if it clearly says you must deposit funds before you can trade. It's not terrible this way either though.. definitely an improvement over how it was before.

martonp commented:
Never mind, you're right, it's better like this. I do notice a few issues though:
1. When I hover over the button, I see a question mark then the tooltip only shows up a few seconds later. Maybe the error message should be the text of the button?
2. When I switch markets, the button is not updated.
3. When I deposit funds, and the insufficient balance error is no longer true, the button should update. 

https://github.com/decred/dcrdex/assets/6186350/b388bfdc-dca0-4060-bbc1-c7ed0e22aff8

peterzen commented:
> Never mind, you're right, it's better like this. I do notice a few issues though:
> 
>     1. When I hover over the button, I see a question mark then the tooltip only shows up a few seconds later. Maybe the error message should be the text of the button?

Fixed the cursor to `not-allowed`, the tooltip behaves correcly now.

> 
>     2. When I switch markets, the button is not updated.
> 
Solved - when switching markets the button changes to a disabled state as the form values are cleared.

>     3. When I deposit funds, and the insufficient balance error is no longer true, the button should update.
> 
The button state updates correctly now, also refactored the update logic to make the changes more readable.

martonp commented:
https://github.com/decred/dcrdex/assets/6186350/6bec2a18-df72-4034-8530-58a486096fee


https://github.com/decred/dcrdex/assets/6186350/8514012f-8453-49f1-9818-9d237ad362dd


For me the tooltip still looks the same. Also, I noticed that the market order section is broken. For market buys, the button is always disabled, even if the balance is enough. The market sells button also has some strange behavior you can see in the videos.

peterzen commented:
> For me the tooltip still looks the same. Also, I noticed that the market order section is broken. For market buys, the button is always disabled, even if the balance is enough. The market sells button also has some strange behavior you can see in the videos.

Thanks for testing this, the bug causing those issues is fixed.  I've also added a test for the rate field being non-zero, and sorted the problem of the UI not always being updated when a new balanceNote came in.

martonp commented:
The logic looks good now. How do you feel about this clean up though? https://github.com/decred/dcrdex/commit/0487795c2ad94e52a9a17f799487d2719940b6f6

Also, the hover over the button still has the same behavior for me. I see that error icon for a bit before the tooltip shows up. How do you feel about just updating the button text?

peterzen commented:
> The logic looks good now. How do you feel about this clean up though? [0487795](https://github.com/decred/dcrdex/commit/0487795c2ad94e52a9a17f799487d2719940b6f6)
> 
Thanks for that, added.

> Also, the hover over the button still has the same behavior for me. I see that error icon for a bit before the tooltip shows up. How do you feel about just updating the button text?

Would it be clearer if the cursor stayed the default, instead of changing to the error icon on hover?

The button might break the layout if we changed the text, translated strings might be longer than the English default and even that is at the limits.

![image](https://github.com/decred/dcrdex/assets/5878500/e6df37c1-956c-4ff4-aa44-6cf84ab5245b)

martonp commented:
> Would it be clearer if the cursor stayed the default, instead of changing to the error icon on hover?

Yeah I think the error icon is kind of confusing. I've never really seen that before.
I guess no need for the button text updating. It's clear enough.

martonp commented:
I think a better UI would be to actually just allow the button, but when you click it, it does a form validation with error message(s) below the invalid fields.

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -1008,6 +1021,71 @@ export default class MarketsPage extends BasePage {
+  }
+
+  updateOrderBttnState () {
+    const mkt = this.market
+    const order = this.currentOrder
+    const orderQty = order.qty
+    const orderRate = order.rate
```
```Typescript
const { market: mkt, currentOrder: { qty: orderQty, rate: orderRate, isLimit, sell } } = this
```

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -2512,13 +2598,10 @@ export default class MarketsPage extends BasePage {
     const avail = note.balance.available
     switch (note.assetID) {
       case mkt.baseCfg.id:
-        // If we're not showing the max order panel yet, don't do anything.
-        if (!mkt.maxSell) break
```
Hmm. Why?

peterzen replied:
This way the max order panel and the order button update when the balance does.

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -2590,12 +2673,19 @@ export default class MarketsPage extends BasePage {
-    page.qtyField.value = String(lots * lotSize / this.market.baseUnitInfo.conventional.conversionFactor)
+    page.qtyField.value = String(orderQty / this.market.baseUnitInfo.conventional.conversionFactor)
+
+    if (!this.isLimit() && this.isSell()) {
+      const baseWallet = app().assets[this.market.base.id].wallet
+      this.setOrderBttnEnabled(orderQty <= baseWallet.balance.available, intl.prep(intl.ID_ORDER_BUTTON_SELL_BALANCE_ERROR))
+    }
```
You're only setting the button status for market sells here? Why not limit orders?

peterzen replied:
Generally the order button state is managed from `preBuy()` and `preSell()`.  The market buy order logic is different and those methods don't get called so that case has to be handled here.

buck54321 replied:
Please leave a good comment then. My guess is we can handle this somewhere more natural though.

buck54321 replied:
I'm having trouble seeing where market sells are treated differently such that the `updateOrderBttnState` isn't called.

peterzen replied:
It's the market buy orders that don't have the lot field - sorry about the confusion above.  Adding a comment.

buck54321 replied:
That doesn't really help me. This little block only runs for market sells. Why? You said

> The market sell order logic is different and those methods don't get called so that case has to be handled here.

Where is the market sell order logic handled different such that the order button state is not updated.

peterzen replied:
`previewQuoteAmt(true)` doesn't call `preSell` which calls `updateOrderBttnState` to update the button, hence the special case here.

https://github.com/peterzen/dcrdex/blob/204a7d0f03221859cb6170b847ab759ddfb2317f/client/webserver/site/src/js/markets.ts#L1334-L1338

    if (!show || !adjusted || !order.qty) {
      page.orderPreview.textContent = ''
      this.drawChartLines()
      return
    }

buck54321 replied:
Oh. It's because `adjusted` in `NaN`.

buck54321 reviewed (changes requested):
Is this right?

https://github.com/decred/dcrdex/assets/6109680/ff31f736-7094-4fcd-9aa1-7bc5268d0f11

peterzen commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -308,7 +308,11 @@ export default class MarketsPage extends BasePage {
         if (!maxSell) return
         page.lotField.value = String(maxSell.swap.lots)
-      } else page.lotField.value = String(this.market.maxBuys[this.adjustedRate()].swap.lots)
+      } else {
+        const maxBuy = this.market.maxBuys[this.adjustedRate()]
+        if (!maxBuy) return
+        page.lotField.value = String(maxBuy.swap.lots)
+      }
```
IIRC this handles the situation when the element is displayed but the estimate is 0 (this was throwing an exception)

peterzen commented:
> Is this right?
> button-viz.mp4

Could you please repost the vid, it won't render:
![image](https://github.com/decred/dcrdex/assets/5878500/b28e0856-6f3c-4e91-92db-dfff782fb365)

buck54321 commented:
Hmm. Well basically, for a limit buy order, the button state was not reactive to changes in the **lots** field.

peterzen commented:
> Hmm. Well basically, for a limit buy order, the button state was not reactive to changes in the **lots** field.

Was there anything logged in the console?  Was the button tooltip referring to no balance or empty order quantity?
I can't reproduce the problem.

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -1008,6 +1021,68 @@ export default class MarketsPage extends BasePage {
+      btn.removeAttribute('disabled')
+      btn.removeAttribute('title')
+    } else {
+      btn.setAttribute('disabled', 'disabled')
```
`btn.setAttribute('disabled', 'true')`

peterzen replied:
Sigh, those HTML4 habits die hard.

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -1008,6 +1021,68 @@ export default class MarketsPage extends BasePage {
+      }
+      if (mkt.maxSell) {
+        this.setOrderBttnEnabled(orderQty <= mkt.maxSell.swap.value, intl.prep(intl.ID_ORDER_BUTTON_SELL_BALANCE_ERROR))
+      }
+      return
```
Seems like we're just giving up here if we don't have a `mkt.maxSell`. Is that OK? I guess we're just expecting the `preSell` to come back soon if we don't already have it?

peterzen replied:
Yes, once `/api/maxsell` comes back the previous branch will execute.

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -1270,12 +1346,15 @@ export default class MarketsPage extends BasePage {
     const baseWallet = app().assets[mkt.base.id].wallet
     if (baseWallet.balance.available < mkt.cfg.lotsize) {
       this.setMaxOrder(null)
+      this.updateOrderBttnState()
```
Looks like you move `this.updateOrderBttnState` into `setMaxOrder` and reduce duplication.

peterzen replied:
It is a slight duplication, but the intention here was to avoid overloading `maxOrder` with `updateOrderBttnState` calls for the sake of readability.

peterzen replied:
Let me know if you're okay with that or I can optimize this

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -2605,18 +2692,22 @@ export default class MarketsPage extends BasePage {
     const page = this.page
-    const order = this.parseOrder()
-    if (order.qty < 0) {
+    const order = this.currentOrder = this.parseOrder()
+    if (order.qty <= 0) {
```
I can no longer enter any value < 1, since it requires first entering `0.`, and this now removes the decimal point.

peterzen replied:
Thanks for spotting this, it's fixed.

peterzen replied:
I'm looking into it

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -1008,6 +1021,68 @@ export default class MarketsPage extends BasePage {
     } else this.page.submitBttn.textContent = intl.prep(intl.ID_SET_BUTTON_BUY, { asset: Doc.shortSymbol(this.market.baseCfg.unitInfo.conventional.unit) })
   }
 
+  setOrderBttnEnabled (isEnabled: boolean, disabledTooltipMsg?: string) {
+    const btn = this.page.submitBttn
```
I recommend putting a line here.

```Javascript
console.error("setOrderBttnEnabled")
```
and fiddling with input values. Try to get this down to running once per action.

peterzen replied:
The above change eliminated superfluous `updateOrderBttnState` calls.

peterzen commented:
[This commit](https://github.com/decred/dcrdex/pull/2534/commits/80d55be0f1452d2275d183a8c510f0192afe7a72) is somewhat unrelated but feels like it belongs here

buck54321 commented on client/webserver/site/src/js/markets.ts:
```diff
@@ -2590,12 +2677,20 @@ export default class MarketsPage extends BasePage {
-    page.qtyField.value = String(lots * lotSize / this.market.baseUnitInfo.conventional.conversionFactor)
+    page.qtyField.value = String(orderQty / this.market.baseUnitInfo.conventional.conversionFactor)
+
+    // no lots on market buy orders
```
This is a given and is unrelated to the question at hand. Can delete.
